### ISSUE TITLE
YOLOv3 low accuracy

### ISSUE THREAD

**Issue by Devin97** [AlexeyAB/darknet#5257]
Hey @AlexeyAB I've trained full yolov3 with pedestrian dataset from OpenImages. [Downloaded using OIDToolkitv4]

Train: 2400
Valid: 600
(Split from same dataset)

Trained for 6000 Iterations, used darknet53.conv.74 pretrained weights
batch = 64
subdivisions = 32
width & height = 416

![chart_yolov3-custom-train](https://user-images.githubusercontent.com/20945989/79637456-728e5300-819d-11ea-865b-91a811323c21.png)

- What can I do to improve accuracy?
- Tried to perform detection on a ~1 min video clip, average FPS was around 16~18. What can I do to improve FPS?

**Comment by AlexeyAB** [comment]
> What can I do to improve accuracy?

1. Use more training images
2. Train this cfg https://drive.google.com/open?id=15WhN7W8UZo7-4a0iLkx11Z7_sDVHU4l1 with this pre-trained weights file https://drive.google.com/open?id=1ULnPnamS5A6lOgidlBXD24IdxoDAFaaV and train with flag `-clear` at the end of training command

> Tried to perform detection on a 1 min video clip, average FPS was around 1618. What can I do to improve FPS?

3. Download the latest Darknet version, and recompile with `GPU=1 CUDNN=1 CUDNN_HALF=1 OPENCV=1`

4. What GPU do you use?

**Comment by Devin97** [comment]
Thanks for replying. I'll train again with that config.
 
Im using NVIDIA Tesla T4 (1 GPU) on google cloud.

**Comment by Devin97** [comment]
Also, is there any documentation for all the hyper parameters?

**Comment by AlexeyAB** [comment]
@Devin97 
Read: https://github.com/AlexeyAB/darknet/wiki
Some will be added later.

**Comment by Devin97** [comment]
![SharedScreenshot](https://user-images.githubusercontent.com/20945989/79674180-32ab8800-81fe-11ea-86a4-e03a0c722c6a.jpg)

Hey @AlexeyAB 

After 1000 iterations, mAP is 0% and avg loss is -nan. Is this common?

**Comment by AlexeyAB** [comment]
Something goes wrong.

* Did you change burn_in= or learninig_rate= in cfg-file? Don't do it.

* Try to set before line `##########################` in cfg-file:
```
stopbackward=2000
train_only_bn=1
```

**Comment by Devin97** [comment]
> Did you change burn_in= or learninig_rate= in cfg-file? Don't do it.
No haven't changed these parameters. Just changed batch=64, subdivision=16 and classes=1

Also, is it ok to directly use the full pretrained weights? Do I have to get a portion of these weights using "darknet partial" command?

**Comment by AlexeyAB** [comment]
Better if you will use partial:
https://github.com/AlexeyAB/darknet/blob/4786d557f9ac7452ded74ce2f3df06fbe991924e/build/darknet/x64/partial.cmd#L12

**Comment by Devin97** [comment]
> Better if you will use partial:
> https://github.com/AlexeyAB/darknet/blob/4786d557f9ac7452ded74ce2f3df06fbe991924e/build/darknet/x64/partial.cmd#L12

Getting Segmentation Fault error.
![SharedScreenshot1](https://user-images.githubusercontent.com/20945989/79684674-f6edde00-8250-11ea-9d73-b8c71a5fca07.jpg)

Tried to train with final weights and added these parameters as you suggested
`stopbackward=2000`
`train_only_bn=1`

Still facing the same issue

![SharedScreenshot2](https://user-images.githubusercontent.com/20945989/79684721-35839880-8251-11ea-8142-2ddcd1078f34.jpg)

![chart_cd53paspp-gamma-train](https://user-images.githubusercontent.com/20945989/79684734-43391e00-8251-11ea-93dd-e60d3475d2b4.png)

Here's my cfg file

[cd53paspp-gamma-train.txt](https://github.com/AlexeyAB/darknet/files/4498752/cd53paspp-gamma-train.txt)

**Comment by AlexeyAB** [comment]
1. Don' use `obj.data` file for `partial` command.
`./darknet partial cfg/cd53paspp-omega.cfg cd53paspp-omega_final.weights cd53paspp-omega.conv.137 137 `

2. Use
```
stopbackward=2000
train_only_bn=1
```

3. set `learning_rate=0.001`

**Comment by Devin97** [comment]
I see, I'll train again with learning_rate=0.001
What is the reason for the avg loss to be nan? Does this mean that something's wrong with the dataset?

**Comment by AlexeyAB** [comment]
Exploding gradients(backward) / features(forward): https://machinelearningmastery.com/exploding-gradients-in-neural-networks/

To solve this:
* reduce learning_rate=
* reduce batch=
* increase burn_in=
* increase decay=
* use `max_delta=3` in `[yolo]` layers
* use `stopbackward=2000` and `train_only_bn=1` at the last backbone layer
* use less layers
* fix your dataset
* use another model
* use gradinent clipping

**Comment by Devin97** [comment]
Hey @AlexeyAB 
It's been more than 2000 iterations and I didn't get any "nan" avg loss, so thats good news. However, there's another problem..

![chart_cd53paspp-gamma-train](https://user-images.githubusercontent.com/20945989/79697874-d9496480-82a2-11ea-88c6-79e66a4bd087.png)

Avg loss seems to be not decreasing much. Is it normal for this to happen at the beginning of the training with this cd53paspp config? or something's wrong?

![Screenshot from 2020-04-20 01-06-15](https://user-images.githubusercontent.com/20945989/79697952-3d6c2880-82a3-11ea-83bf-5536f55c5a87.png)

**Comment by AlexeyAB** [comment]
avg loss for cd53paspp  is higher than for yolov3, but the mAP is also higher.

Also the lower learning_rate= - the more stable training - but training is slower. So you should find optimal learning rate.

*  you can keep learninig_rate=0.001 and increase max_batches= and steps= to train longer
* or you can try to add max_delta=3 in each [yolo] layer, and use higher learninig_rate=0.00261

**Comment by Devin97** [comment]
My current configurations has 
learning_rate = 0.001

and I've added max_delta = 3 in yolo layers.

> you can keep learninig_rate=0.001 and increase max_batches= and steps= to train longer
I'll also try this out on next training session.

I'll post my results when the training is finished. Thanks for the help @AlexeyAB !

**Comment by Devin97** [comment]
This is my training results
![chart_cd53paspp-gamma-train (1)](https://user-images.githubusercontent.com/20945989/79741059-39312100-831e-11ea-8d89-98d517fdc538.png)

Accuracy didn't improve, its almost close to yolov3 config. 

Performed detection on a 1min clip. The FPS is aroun 9, but detection is somewhat more stable as compared to yolov3.

Video: https://drive.google.com/open?id=1mMe-S2XL2InaTjhzEQAc3O50fpyPRxlF

**Comment by AlexeyAB** [comment]
Try to train again:
with `stopbackward=2000`
but **without** `train_only_bn=1`

use `max_delta=3` in [yolo] layers.
and use `learning_rate=0.00261`

And set 2x higher max_batches=12000 and steps=...

----

If the avg loss Nan will occur, then train with `learning_rate=0.001`

**Comment by Jureong** [comment]
How did you let mAP show on that chart?
for me, there's only loss.

**Comment by nanhui69** [comment]
> > What can I do to improve accuracy?
> 
> 1. Use more training images
> 2. Train this cfg https://drive.google.com/open?id=15WhN7W8UZo7-4a0iLkx11Z7_sDVHU4l1 with this pre-trained weights file https://drive.google.com/open?id=1ULnPnamS5A6lOgidlBXD24IdxoDAFaaV and train with flag `-clear` at the end of training command
> 
> > Tried to perform detection on a 1 min video clip, average FPS was around 1618. What can I do to improve FPS?
> 
> 1. Download the latest Darknet version, and recompile with `GPU=1 CUDNN=1 CUDNN_HALF=1 OPENCV=1`
> 2. What GPU do you use?

@AlexeyAB  whats the -clear mean? and is this flag necessary ？ts

**Comment by Devin-botsync** [comment]
> How did you let mAP show on that chart?
> for me, there's only loss.

use -map flag

**Comment by Devin97** [comment]
> How did you let mAP show on that chart?
> for me, there's only loss.

Use -map flag
